features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/ecp_alt.c

```diff
@@ -0,0 +1,2784 @@
+     * (see costs of the various parts, with 1S = 1M)
+     */
+    w = grp->nbits >= 384 ? 5 : 4;
+
+    /*
+     * If P == G, pre-compute a bit more, since this may be re-used later.
+     * Just adding one avoids upping the cost of the first mul too much,
+     * and the memory cost too.
+     */
+    if( p_eq_g )
+        w++;
+
+    /*
+     * Make sure w is within bounds.
+     * (The last test is useful only for very small curves in the test suite.)
+     */
+    if( w > MBEDTLS_ECP_WINDOW_SIZE )
+        w = MBEDTLS_ECP_WINDOW_SIZE;
+    if( w >= grp->nbits )
+        w = 2;
+
+    return( w );
+}
+
+/*
+ * Multiplication using the comb method - for curves in short Weierstrass form
+ *
+ * This function is mainly responsible for administrative work:
+ * - managing the restart context if enabled
+ * - managing the table of precomputed points (passed between the below two
+ *   functions): allocation, computation, ownership tranfer, freeing.
+ *
+ * It delegates the actual arithmetic work to:
+ *      ecp_precompute_comb() and ecp_mul_comb_with_precomp()
+ *
+ * See comments on ecp_comb_recode_core() regarding the computation strategy.
+ */
+/*
+ * Multiplication using the comb method,
+ * for curves in short Weierstrass form
+ *
+ * CYPRESS EDITION
+ */
+static int ecp_mul_comb( mbedtls_ecp_group *grp, mbedtls_ecp_point *R,
+                         const mbedtls_mpi *m, const mbedtls_ecp_point *P,
+                         int (*f_rng)(void *, unsigned char *, size_t),
+                         void *p_rng,
+                         mbedtls_ecp_restart_ctx *rs_ctx )
+{
+
+    #define ciL                 (sizeof(mbedtls_mpi_uint))       /* chars in limb  */
+    #define biL                 (ciL << 3)                       /* bits  in limb  */
+    #define BITS_TO_LIMBS(i)    ((i) / biL + ((i) % biL != 0))
+    uint32_t data_size = BITS_TO_LIMBS(grp->pbits);
+
+    int ret;
+
+    (void)f_rng;
+    (void)p_rng;
+
+    mbedtls_mpi di;
+    mbedtls_ecp_point Pi;
+
+    mbedtls_mpi_init(&di);
+    MBEDTLS_MPI_CHK( mbedtls_mpi_copy(&di, m) );
+
+    mbedtls_ecp_point_init(&Pi);
+    MBEDTLS_MPI_CHK( mbedtls_ecp_copy( &Pi, P ) );
+
+    /* we need N to be odd to transform m in an odd number, check now */
+    if( mbedtls_mpi_get_bit( &grp->N, 0 ) != 1 )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    MBEDTLS_MPI_CHK( mbedtls_mpi_grow(&di, data_size) );
+
+    MBEDTLS_MPI_CHK( mbedtls_mpi_grow(&Pi.X, data_size) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_grow(&Pi.Y, data_size) );
+
+    MBEDTLS_MPI_CHK( mbedtls_mpi_grow(&R->X, data_size) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_grow(&R->Y, data_size) );
+
+    Cy_Crypto_Core_EC_NistP_PointMultiplication (CRYPTO,
+    		CY_CRYPTO_ECC_ECP_SECP256R1,
+			(uint8_t *)Pi.X.p, (uint8_t *)Pi.Y.p,
+			(uint8_t *)di.p,
+			(uint8_t *)R->X.p, (uint8_t *)R->Y.p);
+
+    /* R.Z coordinate should be 1 */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_lset( &R->Z, 1 ) );
```

Comment:
Will response later, need some internal discussion.

Reply:
This means that if the result is the point (1,1,1) in normalized Jacobian coordinates then `Cy_Crypto_Core_EC_NistP_PointMultiplication()` returns (1,1). What does the function do if the result is the point at infinity ((1,1,0) in Jacobian coordinates)?

Reply:
Cy_Crypto_Core_EC_NistP_PointMultiplication() converts a point from Affine to Jacobian coordinates (Z=1) before an operations.
The result of operation converts from Jacobian projective coordinates to affine coordinates by Cy_Crypto_Core_JacobianInvTransform() before return:

/*******************************************************************************
* Function Name: Cy_Crypto_Core_JacobianInvTransform
****************************************************************************//**
*
* Transformation from Jacobian projective coordinates to affine coordinates in GF(VR_P).
* (s_x, s_y, s_z) -> (p_x, p_y), where p_x = s_x/s_z^2, p_y = s_y/s_z^3
********************************************************************************

What we should do to make a proper result data?

Reply:
You can pass parameters to `Cy_Crypto_Core_EC_NistP_PointMultiplication()` such that the result is (1,1,0) in Jacobian coordinates. This point does not have a representation in affine coordinates (you would have to divide by zero in the conversion).

Since `Cy_Crypto_Core_EC_NistP_PointMultiplication()` represents the result in affine coordinates, the only sensible thing to do that comes to mind is to return an error. Then the caller could detect this error and set the point `R` to (1,1,0) manually .